### ISSUE TITLE
Rule Change: no-constant-binary-expression should report `<expr1> ?? <non-nullish> ?? <expr2>`

### ISSUE THREAD

**Issue by hornta** [eslint/eslint#16408]
### Rule details

Inform developer of potentially unnecessary nullish coalescing operator

### Related ECMAScript feature

Nullish coalesing operator

### What type of rule is this?

Suggests an alternate way of doing something

### Example code

```js
// fail
window.abc ?? "any non-nullish value" ?? "any non-nullish value" // warn (the last nullish coalesing operator will never be used)
window.abc ?? undefined ?? "any non-nullish value" // warn (the first nullish coalescing operator is skipped)
window.abc ?? null ?? "any non-nullish value" // warn (the first nullish coalescing operator is skipped)
```

### Why should this rule be in the core instead of a plugin?

I don't know.

### Participation

- [ ] I am willing to submit a pull request to implement this rule.

**Comment by mdjermanovic** [comment]
Hi @hornta, thanks for the issue!

> ```js
> window.abc ?? "any non-nullish value" ?? "any non-nullish value" // warn (the last nullish coalesing operator will never be used)
> ```

I think this would fit into the existing [no-constant-binary-expression](https://eslint.org/docs/latest/rules/no-constant-binary-expression) rule, because the second `??` always short-circuits.

```js
/* eslint no-constant-binary-expression: error */

window.abc && false && anything; // error, because window.abc && false is always falsy
window.abc || true || anything; // error, because window.abc || true is always truthy
window.abc ?? "non-nullish" ?? anything; // could be error, because window.abc ?? "non-nullish" is always non-nullish
```

[playground demo](https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLWNvbnN0YW50LWJpbmFyeS1leHByZXNzaW9uOiBlcnJvciAqL1xuXG53aW5kb3cuYWJjICYmIGZhbHNlICYmIGFueXRoaW5nOyAvLyBlcnJvciwgYmVjYXVzZSB3aW5kb3cuYWJjICYmIGZhbHNlIGlzIGFsd2F5cyBmYWxzeVxud2luZG93LmFiYyB8fCB0cnVlIHx8IGFueXRoaW5nOyAvLyBlcnJvciwgYmVjYXVzZSB3aW5kb3cuYWJjIHx8IHRydWUgaXMgYWx3YXlzIHRydXRoeVxud2luZG93LmFiYyA/PyBcIm5vbi1udWxsaXNoXCIgPz8gYW55dGhpbmc7IC8vIGNvdWxkIGJlIGVycm9yLCBiZWNhdXNlIHdpbmRvdy5hYmMgPz8gXCJub24tbnVsbGlzaFwiIGlzIGFsd2F5cyBub24tbnVsbGlzaFxuIiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoibGF0ZXN0Iiwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6eyJlczYiOnRydWV9fX0=)

> ```js
> window.abc ?? undefined ?? "any non-nullish value" // warn (the first nullish coalescing operator is skipped)
> window.abc ?? null ?? "any non-nullish value" // warn (the first nullish coalescing operator is skipped)
> ```

In these cases, neither of the two `??` operators neither always nor never short-circuits, so this would be a separate rule. But, this check can be relatively easy configured in [no-restricted-syntax](https://eslint.org/docs/latest/rules/no-restricted-syntax), so I'm not sure if we want to add another rule.

```js
/* eslint no-restricted-syntax: ["error", {
    "selector": "LogicalExpression[operator='??'] > LogicalExpression[operator='??'].left > :matches(Literal[value=null], Identifier[name='undefined']).right",
    "message": "Unnecessary nullish coalescing."
}] */

window.abc ?? undefined ?? "any non-nullish value" // error
window.abc ?? null ?? "any non-nullish value" // error
```

[playground link](https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXJlc3RyaWN0ZWQtc3ludGF4OiBbXCJlcnJvclwiLCB7XG4gICAgXCJzZWxlY3RvclwiOiBcIkxvZ2ljYWxFeHByZXNzaW9uW29wZXJhdG9yPSc/PyddID4gTG9naWNhbEV4cHJlc3Npb25bb3BlcmF0b3I9Jz8/J10ubGVmdCA+IDptYXRjaGVzKExpdGVyYWxbdmFsdWU9bnVsbF0sIElkZW50aWZpZXJbbmFtZT0ndW5kZWZpbmVkJ10pLnJpZ2h0XCIsXG4gICAgXCJtZXNzYWdlXCI6IFwiVW5uZWNlc3NhcnkgbnVsbGlzaCBjb2FsZXNjaW5nLlwiXG59XSAqL1xuXG53aW5kb3cuYWJjID8/IHVuZGVmaW5lZCA/PyBcImFueSBub24tbnVsbGlzaCB2YWx1ZVwiIC8vIHdhcm4gKHRoZSBmaXJzdCBudWxsaXNoIGNvYWxlc2Npbmcgb3BlcmF0b3IgaXMgc2tpcHBlZClcbndpbmRvdy5hYmMgPz8gbnVsbCA/PyBcImFueSBub24tbnVsbGlzaCB2YWx1ZVwiIC8vIHdhcm4gKHRoZSBmaXJzdCBudWxsaXNoIGNvYWxlc2Npbmcgb3BlcmF0b3IgaXMgc2tpcHBlZCkiLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOiJsYXRlc3QiLCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7fSwiZW52Ijp7ImVzNiI6dHJ1ZX19fQ==)

**Comment by nzakas** [comment]
Yeah, I don’t think it makes sense to add a new rule here.

**Comment by mdjermanovic** [comment]
We seem to have a consensus not to add a new rule.

I'm in favor of updating `no-constant-binary-expression` to report `<expr1> ?? <non-nullish> ?? <expr2>` as explained in https://github.com/eslint/eslint/issues/16408#issuecomment-1275003088. @nzakas what do you think?

**Comment by amareshsm** [comment]
In case we are updating the `no-constant-binary-expression` rule to report an unnecessary nullish coalescing operator then I will work this.

**Comment by nzakas** [comment]
> I'm in favor of updating no-constant-binary-expression to report <expr1> ?? <non-nullish> ?? <expr2> as explained in https://github.com/eslint/eslint/issues/16408#issuecomment-1275003088. @nzakas what do you think?

Makes sense to me. :+1:

**Comment by mdjermanovic** [comment]
Updated the title and marked accepted. @amareshsm this is now ready to implement.
